src/USER-CONP/fix_charge_update.cpp

```diff
@@ -377,23 +395,144 @@ std::vector<int> FixChargeUpdate::local_to_matrix() {
 /* ---------------------------------------------------------------------- */
 
 void FixChargeUpdate::pre_force(int) {
+  update_charges();
+  short_range_correction(true);
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixChargeUpdate::update_charges() {
+  int *mask = atom->mask;
   std::vector<int> mpos = local_to_matrix();
-  int const nmax = atom->nmax;
+  int const nall = atom->nlocal + atom->nghost;
   vector_compute->compute_vector();
   double *b = vector_compute->vector;
-  for (int i = 0; i < nmax; i++) {
-    int const pos = mpos[i];
-    if (pos < 0) continue;
+  for (int i = 0; i < nall; i++) {
+    if (!(groupbit & mask[i])) continue;
     double q_tmp = 0;
     for (int j = 0; j < ngroup; j++) {
-      q_tmp += elastance[pos][j] * (psi[j] - b[j]);
+      q_tmp += elastance[mpos[i]][j] * (psi[j] - b[j]);
     }
     atom->q[i] = q_tmp;
   }
 }
 
 /* ---------------------------------------------------------------------- */
 
+double FixChargeUpdate::compute_scalar() {
+  return electrode_energy(local_to_matrix()) + short_range_correction(false);
+}
+
+/* ---------------------------------------------------------------------- */
+
+double FixChargeUpdate::electrode_energy(vector<int> mpos) {
```

Comment:
electrode_energy finde ich etwas zu generisch, wenn es eigentlich die self_interaction und die Energie durch das Potential Psi ist. Vielleicht aufspalten. Generell würde ich es gut finden, wenn wir die vier Beiträge self, kspace (bzw. LR), infinite boundary (bzw. keq0) sowie short (oder SR) immer getrennt nutzen und sie nicht generisch in anderen methoden mischen. Vielleicht können wir hier etwas konsistenter sein?

Reply:
Habe ich in potential_energy() und self_energy() aufgeteilt

Reply:
Was ist eigentlich genau die potential_energy()? Steht die irgendwo im MW Theo PDF?

Reply:
Habe ich da nicht gefunden, aber in coulomb_self.f90